### PR TITLE
[Libble] Added initial Libble.me tracker support #3897

### PR DESCRIPTION
**Pull request by JakeShirley**
This is my first stab at adding a Jackett tracker.  Apologies in advance for the code crimes.

### CODE REVIEW

**Comment by JakeShirley** [comment]
@cadatoiva / @ngosang Mind looking at this?

**Comment by JakeShirley** [comment]
Ping

**Comment by JakeShirley** [comment]
Amazing, thanks a ton @garfield69

**Review by garfield69 — CHANGES_REQUESTED** [review]
fantastic. Testing LGTM.

currently http://127.0.0.1:9117/api/v2.0/indexers/libble/results/torznab/api?t=caps
returns:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<caps>
  <server title="Jackett" />
  <searching>
    <search available="yes" supportedParams="q" />
    <tv-search available="no" supportedParams="q,season,ep" />
    <movie-search available="no" supportedParams="q" />
    <music-search available="no" supportedParams="" />
    <audio-search available="no" supportedParams="" />
    <book-search available="no" supportedParams="q" />
  </searching>
  <categories>
    <category id="100013" name="Music" />
    <category id="100015" name="Music Videos" />
    <category id="3000" name="Audio">
      <subcat id="3010" name="Audio/MP3" />
      <subcat id="3020" name="Audio/Video" />
      <subcat id="3030" name="Audio/Audiobook" />
      <subcat id="3040" name="Audio/Lossless" />
      <subcat id="3050" name="Audio/Other" />
      <subcat id="3060" name="Audio/Foreign" />
    </category>
    <category id="3020" name="Audio/Video" />
  </categories>
</caps>
```
it would be cool if a future update included support for the music-search to that users of the Lidarr app can do queries with the artist and album ;-)

the categories are misaligned.
cat **music** should be 1, **libblemixtapes** 2 and **musicvideos** 7
and support for category filtering via the **&filter_cat[1]=1** **&filter_cat[2]=1** **&filter_cat[7]=1** could then be supported.

**Comment by JakeShirley** [comment]
@garfield69 I resolved the category issue, thanks for pointing that out!  I can look into music queries now as well.

**Comment by JakeShirley** [comment]
Added music search support.  As far as I can tell all of the parameters are the same as `GazelleTracker`.  Perhaps this can be refactored in the future to share more code with it.

File: src/Jackett.Common/Indexers/Libble.cs [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            var searchUrl = SearchUrl;
+
+            var searchParams = new Dictionary<string, string> { };
+            var queryCollection = new NameValueCollection { };
+
+            // Skip search params if its a test, ask for all torrents
+            if (!query.IsTest)
+            {
+                if (!string.IsNullOrWhiteSpace(query.ImdbID))
+                {
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:133]: remove the braces when possible

File: src/Jackett.Common/Indexers/Libble.cs [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            }
+
+            if (query.Album != null)
+            {
+                queryCollection.Add("groupname", query.Album);
+            }
+
+            searchUrl += "?" + queryCollection.GetQueryString();
+
+            var searchPage = await PostDataWithCookiesAndRetry(searchUrl, searchParams, CookieHeader);
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:173]: If you are not doing "anything special" with the cookies you don't need to pass this parameter.

File: src/Jackett.Common/Indexers/Libble.cs [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                {
+                    var parser = new HtmlParser();
+                    var dom = parser.ParseDocument(result.Content);
+                    var warningNode = dom.QuerySelector("#loginform > .warning");
+                    var errorMessage = warningNode?.TextContent.Trim().Replace("\n\t", " ");
+                    throw new ExceptionWithConfigData(errorMessage, configData);
+                });
+
+
+            SaveConfig();
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:113]: I think you don't need this. If the test pass the config will be saved.

File: src/Jackett.Common/Indexers/Libble.cs [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                OnParseError(searchPage.Content, ex);
+            }
+
+            if (!CookieHeader.Trim().Equals(prevCook.Trim()))
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:315]: I think it's better to not play with cookies. Jackett will do it for your.
Try to remove all cookie related code in this PR and test that everything is still working.

**Review by ngosang — CHANGES_REQUESTED** [review]

File: src/Jackett.Common/Indexers/Libble.cs [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+            var releases = new List<ReleaseInfo>();
+            var searchString = query.GetQueryString();
+            var prevCook = CookieHeader + "";
+            var searchUrl = SearchUrl;
+
+            var searchParams = new Dictionary<string, string> { };
+            var queryCollection = new NameValueCollection { };
+
+            // Skip search params if its a test, ask for all torrents
+            if (!query.IsTest)
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:130]: I think you don't need this check. The code inside this condition will work well in regular and test queries.

**Review by ngosang — CHANGES_REQUESTED** [review]

File: src/Jackett.Common/Indexers/Libble.cs (hunk 4 above)
Comment (JakeShirley) [reply src/Jackett.Common/Indexers/Libble.cs:315]: Sounds good, sorry! This was initially copy/pasted from the XSpeeds tracker (as that was another one I had access to and could test).

File: src/Jackett.Common/Indexers/Libble.cs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+            return IndexerConfigurationStatus.RequiresTesting;
+        }
+
+        protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuery query)
+        {
+            // https://libble.me/torrents.php?searchstr=the+used&taglist=&tags_type=1&order_by=time&order_way=desc
+            var releases = new List<ReleaseInfo>();
+            var searchString = query.GetQueryString();
+            var prevCook = CookieHeader + "";
```
Comment (ngosang) [inline src/Jackett.Common/Indexers/Libble.cs:123]: unused var

**Review by ngosang — APPROVED** [review]

**Comment by ngosang** [comment]
@garfield69 the code LGTM.

**Review by garfield69 — APPROVED** [review]

**Comment by garfield69** [comment]
@JakeShirley  Thank you for your contribution to Jackett.

**Pull request merged**